app/engine/condition.scala

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2017 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package engine
+
+import cats.data.WriterT
+import cats.kernel.Semigroup
+import engine.RoutingReason.RoutingReason
+import model.Location
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+sealed trait Expr[Context, Result] {
+  def evaluate[A <: AuditInfo: Semigroup](context: Context): WriterT[Future, AuditInfo, Result]
+}
+
+case class Pure[C](predicate: C => Future[Boolean], routingReason: RoutingReason) extends Condition[C] with WithRoutingReason
+case class And[C](c1: Condition[C], c2: Condition[C]) extends Condition[C]
+case class Or[C](c1: Condition[C], c2: Condition[C]) extends Condition[C]
+case class Not[C](condition: Condition[C]) extends Condition[C]
+case class When[C](condition: Condition[C])
+
+sealed trait Condition[C] extends Expr[C, Boolean] {
+
+  def evaluate[A <: AuditInfo: Semigroup](context: C): ConditionResult = {
```

Comment:
no, the expected behaviour of the AND, OR and NOT conditions are not well defined.  There is a test hole.